Stein Variational Gradient Descent

twiecki opened the pull request:
Based on: https://github.com/DartML/Stein-Variational-Gradient-Descent
Paper: https://arxiv.org/abs/1608.04471

Still running some experiments but seems to work quite well.

Commit:
Refactor Hamiltonian methods into single class

fonnesbeck committed:
Added mode argument to several step methods and advi to allow mode setting (e.g. nanguardmode) for Theano functions

fonnesbeck committed:
Fixed namespace bugs in mode attribute

fonnesbeck committed:
Reverted function in delta_logp to not accept mode argument

twiecki committed:
ENH Add Stein Variational Gradient Descent.

twiecki committed:
ENH Move SVGD to Theano.

twiecki committed:
BUG Move advi_minibatch to use continuous check in models.py

taku-y committed:
ENH KL-weighting

springcoil commented:
Looks good to me so far. I can't wait to dig into the implementation :)

twiecki committed:
DOC Add Gaussian Process example with SVGD. Also add shortcut notebook dir.

twiecki committed:
DOC Add Gaussian Process example with SVGD. Also add shortcut notebook dir.

twiecki committed:
DOC Add a bit more description for cov matrix.

twiecki commented:
Small example using @fonnesbeck's GP: https://github.com/pymc-devs/pymc3/blob/1a0c66b4938f79d685b8df953db49808918e976f/docs/source/notebooks/gaussian_process.ipynb

taku-y commented:
Very good example! I will take a closer look at the code tonight.
Do you have a plan to support mini-batch?

taku-y committed:
Fix bugs

fonnesbeck commented:
Very cool. I wonder if we can parallelize the gradient calculation over the particles? I would think that would speed things up considerably.

fonnesbeck commented on pymc3/variational/svgd.py:
```diff
@@ -0,0 +1,135 @@
+    x0 += np.random.normal(0, jitter, x0.shape)
+
+    theta = tt.dmatrix('theta')
+    theta.tag.test_value = x0
```
Isn't this where Shared variables should be used instead of tensors (like you would the weights of a neural network)? Seems like theta is an internal state.

twiecki replied:
Yes, good point.

fonnesbeck commented on pymc3/variational/svgd.py:
```diff
@@ -0,0 +1,135 @@
+        else:
+            historical_grad = alpha * historical_grad + (1 - alpha) * (grad_theta ** 2)
+        adj_grad = np.divide(grad_theta, fudge_factor+np.sqrt(historical_grad))
+        theta = theta + stepsize * adj_grad
```
Shouldn't the stepsize get smaller as the procedure iterates, or does it not matter?

ferrine committed:
ENH User model (#1525)

* Started to write Base class for pymc3.models

* mode `add_var` to public api

* Added some docstrings

* Added some docstrings

* added getitem and fixed a typo

* added assertion check

* added resolve var method

* decided not to add resolve method

* Added linear component

* Docs fix

* patsy's intercept is inited properly now

* refactored code

* updated docs

* added possibility to init coefficients with random variables

* added glm

* refactored api, fixed formula init

* refactored linear model, extended acceptable types

* moved useful matrix and labels creation to utils file

* code style

* removed redundant evaluation of shape

* refactored resolver for constructing matrix and labels

* changed error message

* changed signature of init

* simplified utils any_to_tensor_and_labels code

* tests for `any_to_tensor_and_labels`

* added docstring for `any_to_tensor_and_labels` util

* forgot to document return type in `any_to_tensor_and_labels`

* refactored code for dict

* dict tests fix(do not check labels there)

* added access to random vars of model

* added a shortcut for all variables so there is a unified way to get them

* added default priors for linear model

* update docs for linear

* refactored UserModel api, made it more similar to pm.Model class

* Lots of refactoring, tests for base class, more plain api design

* deleted unused module variable

* fixed some typos in docstring

* Refactored pm.Model class, now it is ready for inheritance

* Added documentation for Model class

* Small typo in docstring

* nested contains for treedict (needed for add_random_variable)

* More accurate duplicate implementation of treedict/treelist

* refactored treedict/treelist

* changed `__imul__` of treelist

* added `root` property and `isroot` indicator for base model

* protect `parent` and `model` attributes from violation

* travis' python2 did not fail on bad syntax(maybe it's too new), fixed

* decided not to use functools wrapper

 Unfortunately functools wrapper fails
 when decorating built-in methods so I
 need to fix that improper behaviour.
 Some bad but needed tricks were implemented

* Added models package to setup script

* Refactor utils

* Fix some typos in pm.model

ferrine committed:
ENH User model (#1525)

* Started to write Base class for pymc3.models

* mode `add_var` to public api

* Added some docstrings

* Added some docstrings

* added getitem and fixed a typo

* added assertion check

* added resolve var method

* decided not to add resolve method

* Added linear component

* Docs fix

* patsy's intercept is inited properly now

* refactored code

* updated docs

* added possibility to init coefficients with random variables

* added glm

* refactored api, fixed formula init

* refactored linear model, extended acceptable types

* moved useful matrix and labels creation to utils file

* code style

* removed redundant evaluation of shape

* refactored resolver for constructing matrix and labels

* changed error message

* changed signature of init

* simplified utils any_to_tensor_and_labels code

* tests for `any_to_tensor_and_labels`

* added docstring for `any_to_tensor_and_labels` util

* forgot to document return type in `any_to_tensor_and_labels`

* refactored code for dict

* dict tests fix(do not check labels there)

* added access to random vars of model

* added a shortcut for all variables so there is a unified way to get them

* added default priors for linear model

* update docs for linear

* refactored UserModel api, made it more similar to pm.Model class

* Lots of refactoring, tests for base class, more plain api design

* deleted unused module variable

* fixed some typos in docstring

* Refactored pm.Model class, now it is ready for inheritance

* Added documentation for Model class

* Small typo in docstring

* nested contains for treedict (needed for add_random_variable)

* More accurate duplicate implementation of treedict/treelist

* refactored treedict/treelist

* changed `__imul__` of treelist

* added `root` property and `isroot` indicator for base model

* protect `parent` and `model` attributes from violation

* travis' python2 did not fail on bad syntax(maybe it's too new), fixed

* decided not to use functools wrapper

 Unfortunately functools wrapper fails
 when decorating built-in methods so I
 need to fix that improper behaviour.
 Some bad but needed tricks were implemented

* Added models package to setup script

* Refactor utils

* Fix some typos in pm.model

twiecki commented:
@taku-y Yes, I'd like to support mini-batch, should not be too difficult but haven't looked at advi_minibatch enough. Is the local vs global RV distinction important here? How much of the API do you think we could reuse?

taku-y committed:
Remove unnecessary comments

taku-y commented:
@twiecki The distinction between the local and global RVs is related to autoencoding VB, not to mini-batch. Since I don't understand the SVGD code, I'm not sure reusing the advi_minibatch code is good or not good, sorry.

Commit:
Reformat docs

taku-y committed:
Fix typo

taku-y committed:
Minor fixes

taku-y committed:
Check transformed RVs using hasattr

ferrine committed:
shape problem when sampling

ferrine committed:
tests passed

twiecki committed:
ENH Made theta be shared and move more updating into theano.

twiecki commented:
@fonnesbeck Turned `theta` into a shared, which allowed some other simplifications.

> Very cool. I wonder if we can parallelize the gradient calculation over the particles? I would think that would speed things up considerably.

I think it's already doing that, at least it's using all my cores when running it.

taku-y committed:
Update conv-vae notebook

ferrine committed:
refactored GARCH and added Mv(Gaussian/StudentT)RandomWalk (#1603)

* refactored GARCH and added Mv(Gaussian/StudentT)RandomWalk

* refactored garch logp

* added docs

* fix typo

* even more typos

* better description for tau

DartML committed:
Add updates.py, using Adagrad optimizer for SVGD as default

DartML committed:
Add updates.py, using Adagrad optimizer for SVGD as default

DartML committed:
fix a bug, "from .updates import *"

twiecki commented:
OK, thanks to @DartML we now have ADAM as an optimizer here.
Will add tests and docs and I think then we can merge.

ferrine commented:
I'm finishing #1600. What about further refactoring SVGD with this API?

twiecki commented:
Yes, we definitely should. Any pointers?

ferrine commented:
It seems that optimizers cant, take list of shared for computing updates, could you refactor that please? Lasagne updates are good example of such implementation

twiecki commented:
Could you be more specific as to how you'd want to pass those in?

ferrine commented:
I mean this API ([link](https://github.com/Lasagne/Lasagne/blob/master/lasagne/updates.py#L545)) as it is nat always practical to have only one shared variable
```python
# source: lasagne.updates
def adam(loss_or_grads, params, learning_rate=0.001, beta1=0.9,
         beta2=0.999, epsilon=1e-8):
    """
    Parameters
    ----------
    loss_or_grads : symbolic expression or list of expressions
        A scalar loss expression, or a list of gradient expressions
    params : list of shared variables
        The variables to generate update expressions for
    ...
    """
```

fonnesbeck committed:
Generalized Bound to work with discrete variables

fonnesbeck committed:
Fixed distributions/__init__.py to reflect change to Bound location

twiecki committed:
MAINT Replace discrete RV check.

MAINT Replace discrete RV check.

twiecki commented:
This PR is currently broken (as of merging #1612), it does not converge to the correct posterior anymore. Perhaps @DartML has an idea.

twiecki committed:
MAINT Use theano GammaLn and Psi functions. (#1628)

fonnesbeck committed:
Bounded now inherits from Distribution

twiecki committed:
Remove static reference to theano.config.floatX. Use .astype(). Addresses @ferrine's comments.

Remove static reference to theano.config.floatX. Use .astype(). Addresses @ferrine's comments.

twiecki committed:
ENH Add floatX() to theanof and make advi use it.

twiecki committed:
Remove nan_. Add doc-string.

twiecki committed:
ENH default to asarray if int or float is passed.

twiecki committed:
Merge pull request #1627 from pymc-devs/floatX

MAINT Add floatX function and make ADVI use it.

twiecki committed:
BUG Do not index on svgd_grad. Thanks @DartML for fix.

twiecki commented:
Closing in favor of https://github.com/pymc-devs/pymc3/pull/1671.
